src/hdf5/cloudsat_netcdf_to_ioda.py

```diff
@@ -0,0 +1,162 @@
+#!/home/imoradi/packages/anaconda3/bin/python3.4
+# 2023, Isaac Moradi
+
+import sys
+import netCDF4 as nc
+import xarray as xr
+import numpy as np
+import datetime
+import time
+import read_cloudsat
+
+# fname_in = ['2009335004815_19116_CS_2B-GEOPROF_GRANULE_P1_R05_E02_F00.hdf']
+fname_in = ['2009212223327_17338_CS_2B-GEOPROF_GRANULE_P1_R05_E02_F00.hdf']
+fname_out = "cpr_cloudsat_obs_2009335.nc4"
+
+cpr_obs = read_cloudsat.read_cloudsat(fname_in)
+cpr_obs["sequenceNumber"] = xr.DataArray(np.arange(cpr_obs.obs_id.size), cpr_obs.obs_id.coords)
+cpr_obs = cpr_obs.rename_vars({"elevation": "elevation1"})
+cpr_obs = cpr_obs.stack(Location=['obs_id', 'elevation']).reset_index("Location")
+cpr_obs = cpr_obs.transpose("Location", "channel")
+
+locid = (cpr_obs.obs.values < -100) | (cpr_obs.obs.values > 100) | np.isnan(cpr_obs.obs.values) | np.isinf(cpr_obs.obs.values)
+locid = cpr_obs.Location.values[np.sum(locid,axis=1) == 0]
+cpr_obs = cpr_obs.isel(Location=locid)
+
+nobs = cpr_obs.Location.size
+nchan = cpr_obs.channel.size
+records = np.unique(cpr_obs.sequenceNumber.values).flatten()
+
+fill_value = 9.96921e+36
+fill_value_dbz = np.float32(-9999)
+
+refl_dims = ('Location', 'Channel') 
+tmpMat = np.ones((nobs, nchan,), dtype=np.float32)
+
+# Create the netCDF4 dataset
+with nc.Dataset(fname_out, 'w', format='NETCDF4', zlib=True) as ds:
+    ds.createDimension('nrecs', 1)                
+    ds.createDimension('Channel', nchan)
+    ds.createDimension('Location', nobs)
+                       
+    ds.createVariable('Channel', np.int32, ('Channel',))
+    ds.createVariable('Location', np.float32, ('Location',))
+    ds.createVariable('nrecs', np.int32, ('nrecs',))
+                    
+    ds['nrecs'][:] = np.int32(records.size)
+    ds['Location'][:] = cpr_obs.Location.values.astype(np.int32)
+    ds['Channel'][:] = cpr_obs.channel.values.astype(np.int32)
+
+    
+    ds['Location'].setncattr("suggested_chunk_dim", 100)
+    ds['Channel'].setncattr("suggested_chunk_dim", 100)
+    
+    ds.setncattr_string("_ioda_layout", "ObsGroup")
+    ds.setncattr("_ioda_layout_version", np.int32(0))
+    ds.setncattr("date_time", np.int32(2009113003))
+    ds.setncattr("satellite", "cloudsat")
+    ds.setncattr("sensor", "cpr")
+    ds.setncattr('history', 'Created by Isaac Moradi NASA/GMAO Sept 6, 2023')
+
+    # Groups ['GsiHofX', 'MetaData', 'ObsError', 'ObsValue', 'PreQC', 'RecMetaData']
+
+    g = 'ObsValue'
+    ds_g = ds.createGroup(g)
+    zvar =  ds_g.createVariable('ReflectivityAttenuated', np.float32, refl_dims, fill_value=fill_value_dbz, zlib=True)
+    ds_g.variables['ReflectivityAttenuated'][:] = cpr_obs.obs.values.astype(np.float32)
+    zvar.setncattr_string("units", str("dBz"))
+    
+    
+    g = 'GsiHofX'
```

Comment:
These are non-standard groups and do not exist in the observational data.  

More specifically `Gsi` is obviously not generic across partners so this decoder should function just to decode raw data files